From e51b4f83c9c6a01f344ef78eb995293ea3af2a4f Mon Sep 17 00:00:00 2001
From: OpenEmbedded <mario.schuknecht@dresearch-fe.de>
Date: Fri, 17 Jun 2022 14:23:28 +0200
Subject: [PATCH] adv7180: Reset on set power HYP-27448

Extra reset to improve the interaction with MIPI CSI2 interface.

Signed-off-by: OpenEmbedded <mario.schuknecht@dresearch-fe.de>
---
 drivers/media/i2c/adv7180.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/drivers/media/i2c/adv7180.c b/drivers/media/i2c/adv7180.c
index 5a2263d99b72..e9dd5d05f9d8 100644
--- a/drivers/media/i2c/adv7180.c
+++ b/drivers/media/i2c/adv7180.c
@@ -66,6 +66,8 @@
 #define ADV7180_HUE_DEF		0
 #define ADV7180_HUE_MAX		128
 
+#define ADV7180_REG_DEFAULT_VALUE_Y	0x000c
+#define ADV7180_REG_DEFAULT_VALUE_C	0x000d
 #define ADV7180_REG_CTRL		0x000e
 #define ADV7180_CTRL_IRQ_SPACE		0x20
 
@@ -191,6 +193,8 @@ struct adv7180_state;
 #define ADV7180_FLAG_MIPI_CSI2		BIT(2)
 #define ADV7180_FLAG_I2P		BIT(3)
 
+static int adv7180_set_field_mode(struct adv7180_state *state);
+
 struct adv7180_chip_info {
 	unsigned int flags;
 	unsigned int valid_input_mask;
@@ -516,6 +520,21 @@ static int adv7180_set_power(struct adv7180_state *state, bool on)
 
 	if (state->chip_info->flags & ADV7180_FLAG_MIPI_CSI2) {
 		if (on) {
+			/* Extra reset to improve the interaction with MIPI CSI2 interface */
+			int insel = adv7180_read(state, ADV7180_REG_INPUT_CONTROL);
+			adv7180_write(state, 0x0F, 0x80);
+			/*  Reset ADV7280A-M */
+			msleep(10);
+			adv7180_write(state, 0x0F, 0x0);
+			/* Restore input */
+			adv7180_write(state, ADV7180_REG_INPUT_CONTROL, insel);
+			/* Default to black output */
+			adv7180_write(state, ADV7180_REG_DEFAULT_VALUE_Y, 0x02);
+			adv7180_write(state, ADV7180_REG_DEFAULT_VALUE_C, 0x88);
+			/* Reinitialize */
+			state->chip_info->init(state);
+			adv7180_set_field_mode(state);
+
 			adv7180_csi_write(state, 0xDE, 0x02);
 			adv7180_csi_write(state, 0xD2, 0xF7);
 			adv7180_csi_write(state, 0xD8, 0x65);
-- 
2.35.3

